controller/konnect/ops_kongconsumer.go

```diff
@@ -0,0 +1,195 @@
+package konnect
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	sdkkonnectgo "github.com/Kong/sdk-konnect-go"
+	sdkkonnectgocomp "github.com/Kong/sdk-konnect-go/models/components"
+	sdkkonnectgoops "github.com/Kong/sdk-konnect-go/models/operations"
+	"github.com/Kong/sdk-konnect-go/models/sdkerrors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	k8sutils "github.com/kong/gateway-operator/pkg/utils/kubernetes"
+
+	configurationv1 "github.com/kong/kubernetes-configuration/api/configuration/v1"
+	konnectv1alpha1 "github.com/kong/kubernetes-configuration/api/konnect/v1alpha1"
+)
+
+func createConsumer(
+	ctx context.Context,
+	sdk *sdkkonnectgo.SDK,
+	consumer *configurationv1.KongConsumer,
+) error {
+	if consumer.GetControlPlaneID() == "" {
+		return fmt.Errorf("can't create %T %s without a Konnect ControlPlane ID", consumer, client.ObjectKeyFromObject(consumer))
+	}
+
+	resp, err := sdk.Consumers.CreateConsumer(ctx,
+		consumer.Status.Konnect.ControlPlaneID,
+		kongConsumerToSDKConsumerInput(consumer),
+	)
+
+	// TODO: handle already exists
+	// Can't adopt it as it will cause conflicts between the controller
+	// that created that entity and already manages it, hm
+	if errWrapped := wrapErrIfKonnectOpFailed[configurationv1.KongConsumer](err, CreateOp, consumer); errWrapped != nil {
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				KonnectEntityProgrammedConditionType,
+				metav1.ConditionFalse,
+				"FailedToCreate",
+				errWrapped.Error(),
+				consumer.GetGeneration(),
+			),
+			consumer,
+		)
+		return errWrapped
+	}
+
+	consumer.Status.Konnect.SetKonnectID(*resp.Consumer.ID)
+	k8sutils.SetCondition(
+		k8sutils.NewConditionWithGeneration(
+			KonnectEntityProgrammedConditionType,
+			metav1.ConditionTrue,
+			KonnectEntityProgrammedReason,
+			"",
+			consumer.GetGeneration(),
+		),
+		consumer,
+	)
+
+	return nil
+}
+
+func updateConsumer(
+	ctx context.Context,
+	sdk *sdkkonnectgo.SDK,
+	cl client.Client,
+	consumer *configurationv1.KongConsumer,
+) error {
+	// TODO(pmalek) handle other types of CP ref
+	// TODO(pmalek) handle cross namespace refs
+	nnCP := types.NamespacedName{
+		Namespace: consumer.Namespace,
+		Name:      consumer.Spec.ControlPlaneRef.KonnectNamespacedRef.Name,
```

Comment:
I missed this in previous PRs, but dereferencing  `Spec.ControlPlaneRef.KonnectNamespacedRef` has a potential of nil ptr dereference (for `ControlPlaneRef` and `KonnectNamespacedRef`). Do I understand correctly this will be handled for all entities' types in the scope of the TODO above? Should we create a separate issue for this so we do not forget?

Reply:
All these objects should have the cp ref at this point (this is handled by `handleControlPlaneRef`) but just to be sure I refactored it to check it in update funcs.

8d88627c77a4ce89550dd1562a5a23afb141e8bc

Reply:
Ok, now I see it. `kongConsumerRefersToKonnectControlPlane` filter makes sure we do not reconcile entities with no CP ref. Nevermind.